### PR TITLE
[Experimental] Add useInsertionEffect

### PR DESCRIPTION
**Pull request by rickhanlonii**
## Overview

Adds back the experimental `useMutationEffect` [hook](https://github.com/facebook/react/pull/14336) as `useInsertionEffect`, initially intended for stylesheet libraries.

## Semantics

This hook is called right before mutations are made to the host, to allow inserting dependencies of the soon-to-be mutated host nodes. The canonical example use case is for inserting styles into the DOM so that they're available before reading layout information from DOM nodes that they apply to (like reading height in `useLayoutEffect`). Since this hook is limited in scope, this hook does not have access to refs and cannot schedule updates.

The lifecycle of this hook is a little different than other hooks. `useInsertionEffect` will interleave create and destroy while traversing the tree in the beforeMutation phase. This means that as we traverse the tree before mutations (e.g. the same time we call `getSnapshotBeforeUpdate`), and for each component we will destroy all of the insertion effects and then create all of the insertion effects for that component. 

The resulting in ordering is something like:

```
Destroy Insertion 1 for Component A
Destroy Insertion 2 for Component A
Create Insertion 1 for Component A
Create Insertion 2 for Component A
Destroy Insertion 1 for Component B
Destroy Insertion 2 for Component B
Create Insertion 1 for Component B
Create Insertion 2 for Component B
```

Contrast this with layout effects, which do a two passes. One to destroy all of the layout effects in every component, and a second to create them:

``` 
Destroy Layout 1 for Component A
Destroy Layout 2 for Component A
Destroy Layout 1 for Component B
Destroy Layout 2 for Component B
Create Layout 1 for Component A
Create Layout 2 for Component A
Create Layout 1 for Component B
Create Layout 2 for Component B
```

### CODE REVIEW

**Comment by gaearon** [comment]
> The old version used effect lists, so this version includes a naive, and probably buggy implementation using the new effect traversals.

I’d say effect lists were the usual cause of bugs so you have a higher chance of doing it right with the current approach!

**Review by acdlite — COMMENTED** [review]
Looks good overall! Let's discuss the timing of the destroy function before merging

File: packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
Comment (ryota-murakami) [inline packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js:2696]: What does `@gate` annotation and `www` meaning?  
Also I've curious about [@gate FIXME](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-reconciler/src/__tests__/ReactNewContext-test.js#L1085). 🤔

**Comment by windmaomao** [comment]
@rickhanlonii just curious is the point of having this hook is to handle something even earlier than `LayoutEffect`? do you have a common use case to use it? Thank you.

File: packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
Comment (rickhanlonii) [reply packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js:2696]: `@gate` is a custom annotation we have to gate tests to only pass in certain conditions. If the condition fails, we verify that the test fails. `www` is a special build variant that we use for internal testing.

**Comment by rickhanlonii** [comment]
@windmaomao I updated the description with the use case 👍

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:1177]: ?

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberCommitWork.new.js:1177]: This was a note to follow-up to see if this behavior as ok.

As written, during unmount this will interleave destroy functions for insertion and layout effects, as:

```
Destroy Insertion 1 for Component A
Destroy Insertion 2 for Component A
Destroy Layout 1 for Component A
Destroy Layout 2 for Component A
Destroy Insertion 1 for Component B
Destroy Insertion 2 for Component B
Destroy Layout 1 for Component B
Destroy Layout 2 for Component B
```

File: packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
Comment (acdlite) [inline packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js:2987]: Test name is wrong

File: packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
Comment (acdlite) [inline packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js:2749]: The one test case I didn't see but we should probably have is where `getSnapshotBeforeUpdate` can read the old style before `useInsertionEffect` mutates it

**Review by acdlite — APPROVED** [review]
Looks good!

File: packages/react-debug-tools/src/ReactDebugHooks.js
Comment (gaearon) [inline packages/react-debug-tools/src/ReactDebugHooks.js:195]: Didn't we decide to drop the `ion` and just do `useInsertEffect`?

Comment (rickhanlonii) [reply packages/react-debug-tools/src/ReactDebugHooks.js:195]: Yes, and then no. `useInsertEffect` sounds too much like the thing you're inserting is an effect, which I think we noticed when we started talking about it again and thought of it as a verb first (which is probably the natural reading of it).

File: packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
Comment (rickhanlonii) [reply packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js:2809]: Test added 👍

File: packages/react-dom/src/server/ReactPartialRendererHooks.js
Comment (Andarist) [inline packages/react-dom/src/server/ReactPartialRendererHooks.js:402]: How are CSS-in-JS libraries supposed to handle SSR then? If the main use for this hook is stylesheet libraries then I think we need to document a compelling strategy for how those should be implemented "end to end". 

In Emotion we currently conditionally render `<style/>` tags "inline" in the SSRed output and expect our hydration function to move them to `<head/>` (most commonly) before React "sees" them - it's tricky, but it sort of works (at least right now). Is this an acceptable approach?

Especially that last part of this error message seems to be off - given the primary use case for this hook. Style-based components cannot be exclusively rendered on the client because that would almost completely obliviate the sense of SSRing applications using those kinds of libraries.

**Comment by windmaomao** [comment]
> @windmaomao I updated the description with the use case 👍

Thanks @rickhanlonii, now i understand the usage. Basically similar to a layoutEffect, but applied right before the mutation finishes.    so the order is more like    insertEffect -> mutationEffect -> layoutEffect -> ... -> passiveEffect.

File: packages/react-dom/src/server/ReactPartialRendererHooks.js
Comment (rickhanlonii) [reply packages/react-dom/src/server/ReactPartialRendererHooks.js:402]: @Andarist short term they'll insert them into the stream separately, chunk points. @sebmarkbage is writing up a tutorial for how to do that as a working group post. Long term, we have an RFC we're going to publish for how we intend to handle this in an integrated way that works isomorphically.

Comment (Andarist) [reply packages/react-dom/src/server/ReactPartialRendererHooks.js:402]: Interesting, thanks for the comment and I can't wait to read how this will work 👍  

However, that probably means that the last paragraph of my comment is still valid - the error message here seems to be somewhat inaccurate and could throw somebody off. Unless you expect library maintainers to ignore that and "know better". 🤔

Comment (sebmarkbage) [reply packages/react-dom/src/server/ReactPartialRendererHooks.js:402]: It's probably wrong in the same sense as the useLayoutEffect warning being wrong since it can also be done differently on the server to inject through different means and needs to be done conditionally to workaround these issues.

**Pull request merged**
